warc_metadata_sidecar.py

```diff
@@ -0,0 +1,207 @@
+        record_count = 0
+        text_mime = 0
+        non_text = 0
+        non_mime = 0
+        for record in ArchiveIterator(stream):
+            url = record.rec_headers.get_header('WARC-Target-URI')
+            if record.rec_type not in ['response', 'resource']:
+                continue
+            if 'text/dns' in record.rec_headers.get_header('Content-Type'):
+                continue
+            record_count += 1
+            record_date = record.rec_headers.get_header('WARC-Date')
+            warcinfo_id = record.rec_headers.get_header('WARC-Warcinfo-ID')
+            warcrecord_id = record.rec_headers.get('WARC-Record-ID')
+            # define specific warc_headers to include in sidecar
+            warc_dict = {'WARC-Date': record_date, 'WARC-Concurrent-ID': warcrecord_id}
+            if warcinfo_id:
+                warc_dict['WARC-Warcinfo-ID'] = warcinfo_id
+
+            # https://github.com/webrecorder/warcio/issues/64
+            rawPayload = io.BytesIO(record.raw_stream.read())
+            if record.http_headers:
+                recordCopy = ArcWarcRecord(
+                    record.format, record.rec_type,
+                    record.rec_headers, rawPayload,
+                    record.http_headers, record.content_type,
+                    record.length
+                    )
+                decodedPayload = recordCopy.content_stream().read()
+                rawPayload.seek(0)
+            else:
+                decodedPayload = rawPayload
+
+            print(url)
+            find_mime_and_puid(fido, rawPayload, url)
+            result_dict = {'encoding': None}
+
+            if fido.mime:
+                image_mimes = re.compile(r'(jpeg|gif|png)')
+                text_format_mimes = re.compile(r'(text|html|xml)')
+
+                if not image_mimes.search(fido.mime):
+                    result_dict = find_character_set(decodedPayload)
+
+                if text_format_mimes.search(fido.mime):
+                    lang_cld = find_language(decodedPayload)
+                    text_mime += 1
+                    string_payload = '{0} {1}\n{2} {3}\n{4} {5}\n{6} {7}'.format(
+                        mime_title, fido.mime,
+                        puid_title, fido.puid,
+                        charset_title, result_dict,
+                        language_title, lang_cld
+                        )
+                elif result_dict['encoding'] is None:
+                    non_text += 1
+                    string_payload = '{0} {1}\n{2} {3}'.format(
+                        mime_title, fido.mime,
+                        puid_title, fido.puid,
+                        )
+                else:
+                    non_text += 1
+                    string_payload = '{0} {1}\n{2} {3}\n{4} {5}'.format(
+                        mime_title, fido.mime,
+                        puid_title, fido.puid,
+                        charset_title, result_dict,
+                        )
```

Comment:
Thinking we may not need the `elif` and `else` if we move the character set part to only doing for text types. Then I think the `elif` becomes the `else`?

Also, the tests should check both the outcomes of text mimetypes and non-text mimetypes.